core/cli/src/lib.rs

```diff
@@ -713,4 +771,135 @@ mod tests {
 		assert!(is_node_name_valid("www.visit.me").is_err());
 		assert!(is_node_name_valid("email@domain").is_err());
 	}
+
+	#[test]
+	fn test_node_key_config_input_ed25519() {
+		fn secret_input(p: Option<String>) -> error::Result<bool> {
+			let s = network::identity::ed25519::SecretKey::generate();
+			let k = NodeKeyParams {
+				node_key_type: NodeKeyType::Ed25519,
+				node_key: Some(format!("{:x}", H256::from_slice(s.as_ref()))),
+				node_key_file: None
+			};
+			node_key_config(k, &p)
+				.map(|c| match c {
+					NodeKeyConfig::Ed25519(network::Secret::Input(s2)) =>
+						s.as_ref() == s2.as_ref(),
+					_ => false
+				})
+		}
+
+		quickcheck(secret_input as fn(_) -> _);
```

Comment:
What is the reason of using quickcheck here in all the tests?

Reply:
Because it provides for convenient generation of random inputs where necessary as well as shrinking, and because a unit test is just a degenerate case of a property test.

Reply:
You mean random generation of `p`? So the random generation of a path that does not exist and maybe pollutes any machine that executes this test? (Based on the assumption that it generates some file).

Nevertheless, I don't see any reason why we should use quickcheck here.

Reply:
> You mean random generation of `p`? So the random generation of a path that does not exist and maybe pollutes any machine that executes this test? (Based on the assumption that it generates some file).

It doesn't. Part of the property being tested is that the choice of network config path has no influence on the result, which must be  a static `network::Secret::Input` (with the same private key), given these input parameters. If it isn't, the test fails.

> Nevertheless, I don't see any reason why we should use quickcheck here.

What is the scope of "here"? This test in particular? There are certainly reasons for occasionally testing very specific inputs, but to me, every "random" constant in tests which could be arbitrary is a code smell and often does not catch bugs for specific inputs. QuickCheck simply makes the generation of random inputs easy and shrinking is very helpful in case of failure.

I won't try to make an argument for QuickCheck here, however, so if you want the tests to be removed or turned into unit tests, just tell me the scope (i.e. all tests in this PR or in this module?) and the desired "random" inputs and number of iterations that you find sufficient (if more than one is even desired).

Reply:
> It doesn't. Part of the property being tested is that the choice of network config path has no influence on the result, which must be a static `network::Secret::Input` (with the same private key), given these input parameters. If it isn't, the test fails.
> 

This sounds like one test that tests `None` and `Some(_)`. You are using QuickCheck in every test, this is just overkill. I don't see the reason to do that.

> > Nevertheless, I don't see any reason why we should use quickcheck here.
> 
> What is the scope of "here"? This test in particular? There are certainly reasons for occasionally testing very specific inputs, but to me, every "random" constant in tests which could be arbitrary is a code smell and often does not catch bugs for specific inputs. QuickCheck simply makes the generation of random inputs easy and shrinking is very helpful in case of failure.
> 
> I won't try to make an argument for QuickCheck here, however, so if you want the tests to be removed or turned into unit tests, just tell me the scope (i.e. all tests in this PR or in this module?) and the desired "random" inputs and number of iterations that you find sufficient (if more than one is even desired).

Remove it everywhere. There is no test, where you really test different inputs that could change the outcome.

Reply:
> This sounds like one test that tests `None` and `Some(_)`.

And what is `_`? Hopefully at least a randomly generated string or `Some("random")`?
 
> Remove it everywhere. There is no test, where you really test different inputs that could change the outcome.

I don't follow, the whole point of testing a property is that it holds for all inputs of a certain kind, i.e. that the outcome doesn't change for different inputs, regardless of how the function is implemented.

Anyways, your wish is my command.